### ISSUE TITLE
Cloning variables that are used several times

### ISSUE THREAD

**Issue by Ploppz** [dtolnay/quote#98]
When I want to use some variable several times in `quote! {}`, I get an error that the value has been moved and thus cannot be used again. For this reason, I found it necessary to make several clones of variables before the `quote!{}`, as in the following snippet from my applications:

```
    let names = vars.0.iter().map(|x| x.name.clone());
    let names2 = names.clone();
    let names3 = names.clone();

    let types = vars.0.iter().map(|x| x.ty.clone());
    let types2 = types.clone();
    let types3 = types.clone();

    let result = TokenStream::from(quote! {
        pub struct #struct_name {
            #( #names: #types ),*
        }
        impl #struct_name {
            pub fn random_search(max_iter: Option<usize>) -> #struct_name {
                let mut config = #struct_name {
                    #( #names2: #types2::default() ),*
                };

                config
            }
            pub fn grid_search(max_iter: Option<usize>) -> #struct_name {
                #struct_name {
                    #( #names3: #types3::default() ),*
                }
            }
            // fn optimize(#( #names : #types ),*) {}
        }

    });
```

Is there a way to avoid this?

**Comment by Goncalerta** [comment]
When you have an iterator, in order to use it you must consume it, so unfortunately you are not able to do any better than this with `quote`, at least as far as I know.

However, if you want to try [`proc-quote`](https://crates.io/crates/proc-quote) (a procedural macro implementation of `quote`) you should be able to get around this. The API is the same as in `quote` so your code should work with both crates with almost no changes.

`proc-quote` solves this because if you give it a slice or `Vec`, it is able to call `.iter()` instead of `into_iter()` (thus borrowing instead of consuming), so your code would work like this:
```
    // Assuming you have `external crate proc_quote;` and `use proc_quote::quote;`
    let names = vars.0.iter().map(|x| x.name.clone()).collect::<Vec<_>>();
    let types = vars.0.iter().map(|x| x.ty.clone()).collect::<Vec<_>>();

    let result = TokenStream::from(quote! {
        pub struct #struct_name {
            #( #names: #types ),*
        }
        impl #struct_name {
            pub fn random_search(max_iter: Option<usize>) -> #struct_name {
                let mut config = #struct_name {
                    #( #names: #types::default() ),*
                };

                config
            }
            pub fn grid_search(max_iter: Option<usize>) -> #struct_name {
                #struct_name {
                    #( #names: #types::default() ),*
                }
            }
            // fn optimize(#( #names : #types ),*) {}
        }

    });
```

**Comment by survived** [comment]
You may handle this with `quote`:
```rust
    let ref names = vars.0.iter().map(|x| x.name.clone()).collect::<Vec<_>>();
    let ref types = vars.0.iter().map(|x| x.ty.clone()).collect::<Vec<_>>();

    let result = TokenStream::from(quote! {
        pub struct #struct_name {
            #( #names: #types ),*
        }
        impl #struct_name {
            pub fn random_search(max_iter: Option<usize>) -> #struct_name {
                let mut config = #struct_name {
                    #( #names: #types::default() ),*
                };

                config
            }
            pub fn grid_search(max_iter: Option<usize>) -> #struct_name {
                #struct_name {
                    #( #names: #types::default() ),*
                }
            }
            // fn optimize(#( #names : #types ),*) {}
        }

    });
```

Since `name` and `types` are `&Vec<_>` now, `into_iter()` won't move them out

**Comment by dtolnay** [comment]
This has been fixed in #109 and will be in the next release (#124).

**Comment by Ploppz** [comment]
Sorry I forgot to reply - thanks!
